### PR TITLE
Add support for bitwise operators

### PR DESCRIPTION
**Pull request by bmakuh**
Addresses #297

## Todo
- [x] rebase to bring in unboxing operator removal
- [x] tests

### CODE REVIEW

**Review by ospencer — COMMENTED** [review]
Yikes, I checked out your branch to see what error you got and it's definitely one of our least helpful errors. That error just means that there's an ambiguity in the grammar.

In this case, it's because `^` is both an infix _and_ a prefix operator (I knew it would be `^` that bit us!). The parser can't tell which you mean whenever you write `^`. This also made me realize that unboxing is currently a function bound to the identifier `^`. To actually get both to work, we'd need to make it something statement-like which would be a bit out of scope for this PR (and would also be a breaking change), so I think the move is to just leave `^` out for now.

Also, it looks like you just didn't add the new tokens to the lexer:
https://github.com/grain-lang/grain/blob/master/compiler/src/parsing/lexer.mll#L194-L196

Forgetting that one would just give you a regular syntax error, though 🙂

**Comment by ospencer** [comment]
But I'm really glad to see this come through! 😄

**Comment by phated** [comment]
I also previously ran into issues with this operator at https://github.com/grain-lang/grain/issues/183 - I believe we chatted in the past about just dropping that symbol as the unbox operator. It's also not really needed much anymore since almost all mutable data can be handled with `let mut`.

I'll try to land a PR tonight that removes that operator as `unbox` so we can keep it for bitwise operations in this PR!

**Comment by bmakuh** [comment]
Thanks for the tips @ospencer and thanks for the unboxing PR @phated! Learning a lot trying to make this work 😅

**Comment by bmakuh** [comment]
Are we going to have ambiguity in situations like these for right shift operators?
```
mut next: Option<Bucket<k, v>>
```
```
mut buckets: Array<Option<Bucket<k, v>>>
```

**Comment by ospencer** [comment]
Nope, we shouldn't have to worry about that! Speaking of which, though, we should probably stick to having only one token each for the carets... We would totally run into some issues if there was a token for `>>>` versus just matching a token for a single `>` three times.

**Comment by phated** [comment]
@bmakuh feel free to rebase on master now that I landed the removal of `^` as unbox 🍻

**Comment by bmakuh** [comment]
@ospencer this should be ready for review now.

File: compiler/src/parsing/parser.dyp [hunk 1]
```diff
@@ -127,11 +128,17 @@ binop_expr :
...
+  | binop_expr(<=p90) llcaret_op eols?   binop_expr(<p100) { Exp.apply ~loc:(symbol_rloc dyp) (mkid_expr dyp [$2]) [$1; $4] } p100
   | binop_expr(<=p90) rcaret_op eols?    binop_expr(<p90) { Exp.apply ~loc:(symbol_rloc dyp) (mkid_expr dyp [$2]) [$1; $4] } p90
+  | binop_expr(<=p90) rrcaret_op eols?   binop_expr(<p100) { Exp.apply ~loc:(symbol_rloc dyp) (mkid_expr dyp [$2]) [$1; $4] } p100
+  | binop_expr(<=p90) rrrcaret_op eols?  binop_expr(<p100) { Exp.apply ~loc:(symbol_rloc dyp) (mkid_expr dyp [$2]) [$1; $4] } p100
   | binop_expr(<=p90) lesseq_op eols?    binop_expr(<p90) { Exp.apply ~loc:(symbol_rloc dyp) (mkid_expr dyp [$2]) [$1; $4] } p90
   | binop_expr(<=p90) greatereq_op eols? binop_expr(<p90) { Exp.apply ~loc:(symbol_rloc dyp) (mkid_expr dyp [$2]) [$1; $4] } p90
+  | binop_expr(<=p40) amp_op eols?       binop_expr(<p70) { Exp.apply ~loc:(symbol_rloc dyp) (mkid_expr dyp [$2]) [$1; $4] } p70
   | binop_expr(<=p40) ampamp_op eols?    binop_expr(<p40) { Exp.apply ~loc:(symbol_rloc dyp) (mkid_expr dyp [$2]) [$1; $4] } p40
+  | binop_expr(<=p30) pipe_op eols?      binop_expr(<p50) { Exp.apply ~loc:(symbol_rloc dyp) (mkid_expr dyp [$2]) [$1; $4] } p50
```
Comment (ospencer) [inline compiler/src/parsing/parser.dyp:141]: It looks like for these some of the priorities for the operands are off. They should just match the priority of the operation (that's on the far right side).

For example, `binop_expr(<=p90) caret_op eols?     binop_expr(<p60)` should be `binop_expr(<=p60) caret_op eols?     binop_expr(<p60)`.

I don't think we really need tests to make sure all of these priorities hold as long as all of the numbers match up. Was there a reason why you needed to make them something different?

**Review by ospencer — COMMENTED** [review]
Woohoo! I left one comment about the priorities, but otherwise it looks great! We might want to keep the original names of the bit-shifting operators in Pervasives in addition to the new syntax, since otherwise it'd be a breaking change. I'm going to double check with the core team though.

Excellent work!

File: compiler/src/parsing/parser.dyp (hunk 1 above)
Comment (bmakuh) [reply compiler/src/parsing/parser.dyp]: Oops, it was a mistake.

**Comment by ospencer** [comment]
@bmakuh We chatted, and we're just going to have the breaking change. I dropped your fix commit (and thank you for changing it so quickly!) and we're pretty much good to merge.

**Review by ospencer — APPROVED** [review]

**Pull request merged**